### ISSUE TITLE
add support for depending on installed targets

### ISSUE THREAD

**Issue by avsm** [ocaml/dune#1155]
It would be useful to be able to depend on installed targets in other dune files, for example for use in aliases.  One example is in the OCaml Platform repository (which vendors common build tools).  We currently have this in our toplevel dune file:

```
(alias
 (name cli)
 (deps
   vendor/opam-core/src/client/opamMain.exe
   vendor/opam-core/src/tools/opam_installer.exe
   vendor/bun/src/bun.exe
   vendor/craml/bin/main.exe
   vendor/dune-release/bin/main.exe
   vendor/merlin/src/frontend/ocamlmerlin.exe
   vendor/merlin/src/frontend/ocamlmerlin_server.exe
   vendor/odoc/src/odoc/bin/main.exe
   vendor/utop/src/top/utop.bc
 ))
```

but it would be much nicer to be able to address the dependencies via `(bin opam)` to reflect their installed locations.  Trying to use `%{bin:opam}` currently results in `Error: %{bin:..} isn't allowed in this position`.

**Comment by rgrinberg** [comment]
@ejgallego, could you confirm that you've encountered this issue as well?

I think it would be nice if `%{bin:..}` also made sure that the action would have access to this binary via PATH (regardless if it's private or public).

@diml @avsm thoughts on the last point?

**Comment by diml** [comment]
Supporting `%{bin:...}` in `(deps ...)` seems good

**Comment by avsm** [comment]
> I think it would be nice if %{bin:..} also made sure that the action would have access to this binary via PATH (regardless if it's private or public).

I think that would be useful to copy files around indeed...

**Comment by diml** [comment]
BTW, we already extend the `PATH` when executing actions, so all we need to do is support `%{bin:...}` in `(deps ...)`. The binary will naturally be in the `PATH`.

**Comment by ejgallego** [comment]
@rgrinberg indeed, thanks! I think that would be useful for us. Sphinx documentation in Coq depends on `coqtop`, so even if we can depend on the binary and pass it using an env var, it seems cleaner to me to just depend on the installed version and use `PATH`.

This would simplify for example splitting the documentation to its own OPAM package [as it has non-trivial build-deps]

**Comment by rgrinberg** [comment]
> BTW, we already extend the PATH when executing actions, so all we need to do is support %{bin:...} in (deps ...). The binary will naturally be in the PATH.

Is that true? I thought this only worked for public binaries. Also, I would prefer a more fine grained PATH for actions. For example, I think it would be best that we only appended `PATH` with binaries that are specified by the dependencies. Right now, I think we'll have a bunch of other stuff in PATH that the user might not have depended on.

**Comment by diml** [comment]
> Is that true? I thought this only worked for public binaries.

So does `%{bin:...}`. To be clear, the way I understand the feature request is as follow: considering the following project:

- bin/dune:

```
(executable (public_name foo))
```

- test/dune:

```
(alias
  (name runtest)
  (deps %{bin:foo})
  (action (system "foo ...")))
```

It is guaranteed that the action in the second dune file will see the local version of `foo` in its search PATH. Does that match what you have in mind?

> Also, I would prefer a more fine grained PATH for actions. For example, I think it would be best that we only appended PATH with binaries that are specified by the dependencies. Right now, I think we'll have a bunch of other stuff in PATH that the user might not have depended on.

Note that it's not restricted to the search path, currently it's the case for all files. For instance, if a command states that it depends only on files a and b in the current directory, it will still see all the other files that happen that happen to be there when the command is executed.

**Comment by ejgallego** [comment]
Note, I am faking this by inserting a dummy `(run 
prog) ` in the actions

**Comment by rgrinberg** [comment]
Yup, you've understood correctly. I understand that we can't create proper sandboxes for these actions, but I think that we can at least be careful about the local additions to the PATH that we make to dune.

**Comment by diml** [comment]
How would you implement it?

**Comment by rgrinberg** [comment]
A simple approach would be the following:

* Create a special sandbox dir for a rule that contains a `%{bin:..}` dependency
* For every `%{bin:..}` in the dependency list, create a symlink rule of that dependency to that sandbox dir
* Before executing the action, append the sandbox dir to `PATH`.

**Comment by ejgallego** [comment]
Note it would be useful for us not to be able to depend only in `%{bin:...}` targets but also over general installed files, for example, templates used by the documentations tools that have their own install rule.

**Comment by rgrinberg** [comment]
We have `%{lib:<name>:<file>}` for that purpose already. Could you give that a try?

**Comment by ejgallego** [comment]
I get the same `%{lib ...} is not allowed in this position`.

**Comment by diml** [comment]
BTW, this might be relevant to this discussion: https://github.com/ocaml/dune/issues/1185. This + the `(package xxx)` dependency allows you to find all kinds of installed files without hassle.

**Comment by rgrinberg** [comment]
> Another thing that would be useful for us is to be able to depend on (library coq.kernel) in rules.

What are the use case for this? I think it's always preferably to depend on the concrete artifacts when possible.

**Comment by ejgallego** [comment]
> What are the use case for this?

Interaction with legacy make-based targets, that expect `-package coq.toplevel` to be ready before being called.

**Comment by rgrinberg** [comment]
And `(package ..)` is not good enough? In the linked issue, I see that you're trying to port some tests that rely on `-package`. Rather than relying on manual ocamlfind calls, can't you have dune build build those executables? Or must the done be specifically done via findlib?

**Comment by rgrinberg** [comment]
I think the design phase for this feature, has ended. For now, let's just implement the least controversial bits that we agree on. That means we can throw out the sandboxing features for now.

Summary of the concrete plan:

* Add support for `%{bin:..}` dependencies and add them to `PATH` when executing actions (if the binaries aren't public).

* Add support for `%{lib:..}` in the dependencies field.

@ejgallego are those two features satisfactory for you?

@emillon @diml unless, there's disapproval, I will start working on this.

**Comment by diml** [comment]
No objection, however it's not clear to me how `%{bin:...}` is supposed to work. Is the idea that if you write `(deps %{bin:../foo/bar.exe})` then you will have `bar` available in the `PATH` inside the action?

I don't find this particularly intuitive, it would seem simpler to me to just say that an executable is available everywhere via the `PATH` inside a project. That would be more in line with how we deal with public executables.

**Comment by ejgallego** [comment]
> @ejgallego are those two features satisfactory for you?

Sure, thanks, I think that would improve developer workflow quite a bit. The problem is that Coq doesn't know how to locate things that are not installed. So when you want to run a test, you need to depend on the installed `.vo` files, such as `%{lib:coq}/lib/coq/theories/Prelude.vo}`.

I think that any simple solution that allows us to do that would work. For binaries we could do with a very simple `(public_bin foo)` field that ensures that foo was installed in the default PATH target.

We could also solve our problem by depending on the build object themselves, and having something like #1185 available, but maybe the way proposed here is simpler.

**Comment by rgrinberg** [comment]
> No objection, however it's not clear to me how %{bin:...} is supposed to work. Is the idea that if you write (deps %{bin:../foo/bar.exe}) then you will have bar available in the PATH inside the action?

That was the idea.

> I don't find this particularly intuitive, it would seem simpler to me to just say that an executable is available everywhere via the PATH inside a project. That would be more in line with how we deal with public executables.

Note that this isn't flexible enough. It's common enough that you want to use a different executable under the same name in a single project. The `env` stanza is best to make such customizations without repetition. By the way, this is why I quite like the feature that I'm proposing, it is modifying the environment in the most "local" way - on a per rule basis.

If it's the magic of `%{bin:..}` adding the binary to the path that's bothering you, we could make it more explicit. For example, we could have an action like:

```
(ensure-in-path %{bin:..} (progn ...)`
```

However, I don't think the behavior that I'm proposing is so magical. Unless the current behavior with public executables is considered magical.

**Comment by diml** [comment]
Not magical, but not necessarily intuitive either. It seems odd to me that using a percent form would modify the execution context in this way. Using a specific action seems better, however the name should reflect that it is modifying the execution context, for instance `(add-to-path x-y ../foo/x_y.exe (run x-y ...))`. Though I still find this quite heavy if you have to do it for every single rule.

> Note that this isn't flexible enough. It's common enough that you want to use a different executable under the same name in a single project. The env stanza is best to make such customizations without repetition. By the way, this is why I quite like the feature that I'm proposing, it is modifying the environment in the most "local" way - on a per rule basis.

I feel like this is too local though. Doing this via the `env` stanza seems fine, plus it allows using different binaries depending on the profile.

**Comment by rgrinberg** [comment]
@ejgallego would the following work for you:

```
(env
 (_
  (bins (%{project_root}/foo.exe bar.as exe as baz.exe))))
```

And note that action rules under this environment would now be able to use the local binary names for dependencies:

```
(rule (with-stdout-to t1 foo.exe))
(rule (with-stdout-to t2 baz.exe))
```

**Comment by ejgallego** [comment]
Using env looks a bit strange doesn't it? It seems like it would need a bit of duplication but I need to think.

Why does `(rule ... (run coqtop ...))` correctly installs `coqtop` but we can't move it to a `(deps ... )` field?

**Comment by rgrinberg** [comment]
> Using env looks a bit strange doesn't it? It seems like it would need a bit of duplication but I need to think.

It shouldn't be so bad as the env stanza applies to all stanzas in its directory recursively. So if your tests are in 1 dir, that's all you will need.

> Why does (rule ... (run coqtop ...)) correctly installs coqtop but we can't move it to a (deps ... ) field?

That's a separate issue. We'll fix that regardless. Note that `coqtop` will be in PATH only because it's a public executable.  The solution I gave above will work for any binary.

**Comment by ejgallego** [comment]
> It shouldn't be so bad as the env stanza applies to all stanzas in its directory recursively. So if your tests are in 1 dir, that's all you will need.

Still feels to me a bit out of place, imagine I have a rule:
```
(rule
  (...)
  (action (run some-prog-that-needs-foo1))
```
using the env I cannot reflect that `some-prog` requires `foo1`, right? So how does dune know that it has to build and _install_ (this issue is about installable things only, right?) foo1 from the `env`?

**Comment by ejgallego** [comment]
So to be clear our problem really is implicit dependencies on installed targets, [for example Sphinx depending on coqtop] In actions where the name of the executable appears we tend to be OK using their private version.

**Comment by ejgallego** [comment]
> Because of the `env` stanza, dune would make `../x/foo.exe` visible as `foo` in the `PATH`, and `%{bin:foo}` would resolve to this `foo` binary in the `PATH`. If `coqtop` is a public binary, then you don't even need to `env` stanza.

Ah, very cool then, thanks very much! We only depend on public binaries / data files but indeed i can see how this is useful in tests [currently indeed we are installing a couple of test-only binaries for this reason]

[small off tangent, in our root Dune file, we cannot do:
```
(env
  (dev (...))
  (release (...))
  ....
  (_ common_options))
```
as indeed the wildcard doesn't take effect here. Does this merit a bug?]

**Comment by diml** [comment]
Yh, that merit a bug. Currently, the semantic is: use the first matching pattern. But we could allow to have several env stanzas in the same file for this case for instance.
